astropy/units/physical.py

```diff
@@ -352,7 +352,7 @@ def __str__(self):
         return "/".join(self._physical_type_list)
 
     @staticmethod
-    def _dimensionally_compatible_unit(obj):
+    def _dimensionally_compatible_unit(obj: object) -> u.UnitBase | NotImplemented:  # ???
```

Comment:
this is a good place for `typing.overload`.

Reply:
Same comment as for `__eq__` - I think we'd want the type information to tell what gives success, not what gives failure, where for this function failure means getting `NotImplemented`.

Note that I really worry about "this is a good place for `typing.overload` - if we are adding more than annotations, suddenly files can become like  [this](https://github.com/numpy/numpy/blob/0a4b2b83eaf3479f352a7fe5a4b378a169ab48f0/numpy/__init__.pyi#L663-L675)
```
    # Builtin types
    @overload
    def __new__(cls, dtype: type[builtins.bool], align: builtins.bool = ..., copy: builtins.bool = ..., metadata: dict[builtins.str, Any] = ...) -> dtype[np.bool]: ...
    @overload
    def __new__(cls, dtype: type[int], align: builtins.bool = ..., copy: builtins.bool = ..., metadata: dict[builtins.str, Any] = ...) -> dtype[int_]: ...
    @overload
    def __new__(cls, dtype: None | type[float], align: builtins.bool = ..., copy: builtins.bool = ..., metadata: dict[builtins.str, Any] = ...) -> dtype[float64]: ...
    @overload
    def __new__(cls, dtype: type[complex], align: builtins.bool = ..., copy: builtins.bool = ..., metadata: dict[builtins.str, Any] = ...) -> dtype[complex128]: ...
    @overload
    def __new__(cls, dtype: type[builtins.str], align: builtins.bool = ..., copy: builtins.bool = ..., metadata: dict[builtins.str, Any] = ...) -> dtype[str_]: ...
    @overload
    def __new__(cls, dtype: type[bytes], align: builtins.bool = ..., copy: builtins.bool = ..., metadata: dict[builtins.str, Any] = ...) -> dtype[bytes_]: ...
```

Can we pretty please agree that `overload` has no place in regular python code files?